### PR TITLE
firebase-dataconnect merge into main branch

### PR DESCRIPTION
**Pull request by dconeybe**
Includes contributions by @cherylEnkidu

### CODE REVIEW

**Comment by dconeybe** [comment]
@rlazo @daymxn The main files that need reviewing, IMO, are:

- build.gradle
- firebase-dataconnect/firebase-dataconnect.gradle.kts
- gradle/libs.versions.toml
- settings.gradle
- subprojects.cfg

File: firebase-dataconnect/firebase-dataconnect.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+  androidTestImplementation(libs.turbine)
+}
+
+tasks.withType<KotlinCompile>().all {
+  kotlinOptions {
+    freeCompilerArgs = listOf("-opt-in=kotlin.RequiresOptIn")
+  }
+}
+
+extra["packageName"] = "com.google.firebase.dataconnect"
```
Comment (daymxn) [inline firebase-dataconnect/firebase-dataconnect.gradle.kts:183]: Do you need this?

File: firebase-dataconnect/firebase-dataconnect.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+  cacheRoot.set(cacheRootDirectory)
+}
+
+// Enable Kotlin "Explicit API Mode". This causes the Kotlin compiler to fail if any
+// classes, methods, or properties have implicit `public` visibility. This check helps
+// avoid  accidentally leaking elements into the public API, requiring that any public
+// element be explicitly declared as `public`.
+// https://github.com/Kotlin/KEEP/blob/master/proposals/explicit-api-mode.md
+// https://chao2zhang.medium.com/explicit-api-mode-for-kotlin-on-android-b8264fdd76d1
+tasks.withType<KotlinCompile>().all {
```
Comment (daymxn) [inline firebase-dataconnect/firebase-dataconnect.gradle.kts:198]: oo I didn't know this was a setting. tbh, this might be a good thing to enable globally for Kotlin SDKs. Possible future work (for investigation) @rlazo

File: firebase-dataconnect/gradleplugin/gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -0,0 +1,7 @@
+distributionBase=GRADLE_USER_HOME
```
Comment (daymxn) [inline firebase-dataconnect/gradleplugin/gradle/wrapper/gradle-wrapper.properties:1]: Why are you using your own wrapper for the included build?

**Review by daymxn — CHANGES_REQUESTED** [review]
Haven't looked through the source code much, just the build files primarily.

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -1,20 +1,24 @@
 [versions]
 # javalite, protoc and protobufjavautil versions should be in sync while updating and
-# it needs to match the protobuf version which grpc has transitive dependency on.
+# it needs to match the protobuf version which grpc has transitive dependency on, which
+# needs to match the version of grpc that grpc-kotlin has a transitive dependency on.
 android-lint = "31.3.2"
 autovalue = "1.10.1"
-coroutines = "1.6.4"
+coroutines = "1.7.3"
```
Comment (daymxn) [inline gradle/libs.versions.toml:7]: We need to look at the test failures and make sure none of them come from this change. I doubt it, but just to sanity check.

File: firebase-dataconnect/firebase-dataconnect.gradle.kts (hunk 1 above)
Comment (dconeybe) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts]: Good point. I've now removed the `extra["packageName"]` definition; I see it's not longer needed as of https://github.com/firebase/firebase-android-sdk/pull/5895.

File: firebase-dataconnect/gradleplugin/gradle/wrapper/gradle-wrapper.properties (hunk 3 above)
Comment (dconeybe) [reply firebase-dataconnect/gradleplugin/gradle/wrapper/gradle-wrapper.properties:1]: Ahh you're right that this extra copy of the gradle wrapper isn't required. I had never tried it without the gradle wrapp and assumed it was needed. I've since removed it.

File: gradle/libs.versions.toml (hunk 4 above)
Comment (dconeybe) [reply gradle/libs.versions.toml:7]: All of the GitHub Actions failures looked to me to just be flakes, unrelated to the coroutines version. There is one persistent GitHub Action that fails, but it's just the "CLA" one, and it's complaining because the PR is so big. So we should be good.

**Review by dconeybe — COMMENTED** [review]
Thank you @daymxn for the thorough review. I've addressed some of your comments and left a few questions that I'd like your opinion on.

File: firebase-dataconnect/firebase-dataconnect.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import io.javalin.Javalin
+import io.javalin.http.staticfiles.Location
+
+plugins {
+  id("firebase-library")
+  id("kotlin-android")
+  id("com.google.protobuf")
+  id("copy-google-services")
+  alias(libs.plugins.dokka)
```
[the remark this answers is not in the document]
Comment (rlazo) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts:27]: +1 to use dackka. 

For your first use case, Android Studio has a "documentation view" that shows a side panel with the docs generated HTML. It only works one method / property at the time, so for quick checks is enough but not for larger changes.

As for the second use case, Android Studio doesn't help :-)

**Review by rlazo — CHANGES_REQUESTED** [review]

File: firebase-dataconnect/gradleplugin/gradle/libs.versions.toml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+[versions]
+androidGradlePlugin = "8.2.1"
+kotlin = "1.9.25"
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/gradleplugin/gradle/libs.versions.toml]: I'm not sure exactly how the gradleplugin's kotlin version of 1.9.25 interops with buildSrc's version 1.8.22. But it works somehow. To be safe, though, I've downgraded the Kotlin version to 1.8.22 to match buildSrc: https://github.com/firebase/firebase-android-sdk/pull/6290/commits/72ae3737aea240f5242ee40f1f5dca01b0b9dbc7

File: firebase-dataconnect/api.txt [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+    method @Nullable public suspend Object suspendingClose(@NonNull kotlin.coroutines.Continuation<? super kotlin.Unit>);
+    method @NonNull public String toString();
+    method public void useEmulator(@NonNull String host = "10.0.2.2", int port = 9399);
+    property @NonNull public abstract com.google.firebase.FirebaseApp app;
+    property @NonNull public abstract com.google.firebase.dataconnect.ConnectorConfig config;
+    property @NonNull public abstract com.google.firebase.dataconnect.DataConnectSettings settings;
+    field @NonNull public static final com.google.firebase.dataconnect.FirebaseDataConnect.Companion Companion;
+  }
+
+  public static final class FirebaseDataConnect.Companion {
```
[the remark this answers is not in the document]
Comment (daymxn) [reply firebase-dataconnect/api.txt:76]: I wonder if `@hide`ing it would also apply to the functions it applies... worth trying out. If it does, then it's fine to keep this as is.

File: firebase-dataconnect/api.txt [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    method @NonNull public Object getResult();
+    method public int hashCode();
+    method @NonNull public String toString();
+    property @NonNull public abstract com.google.firebase.dataconnect.QueryRef<Data,Variables> query;
+    property @NonNull public abstract Object result;
+  }
+
+}
+
+package com.google.firebase.dataconnect.core {
```
[the remark this answers is not in the document]
Comment (daymxn) [reply firebase-dataconnect/api.txt:219]: We generally use `@hide` to hide them from public API stuff. Although we don't enforce `internal` as public API- some of the tooling is using java infra, so annotating with `@hide` helps in those situations.

File: firebase-dataconnect/connectors/connectors.gradle.kts [hunk 9]
```diff
@@ -0,0 +1,113 @@
+/*
```
[the remark this answers is not in the document]
Comment (daymxn) [reply firebase-dataconnect/connectors/connectors.gradle.kts:1]: Ah I like that idea. Alright SGTM!

File: firebase-dataconnect/firebase-dataconnect.gradle.kts (hunk 5 above)
[the remark this answers is not in the document]
Comment (daymxn) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts]: > Thanks for pointing out the DackkaPlugin. You are right that I'm just using it to "preview" what the generated docs would look like, largely to make sure that my markdown looks reasonable when rendered as html and that I didn't accidentally mark something as "public". If Dackka had similar functionality, then I'd definitely use that instead.

When creating a PR, if you make a change that impacts the Javadocs you'll get a [comment like this](https://github.com/firebase/firebase-android-sdk/pull/6292#issuecomment-2364994601) that shows the diff.

If you make a change that impacts the API you'll also get a [comment like this](https://github.com/firebase/firebase-android-sdk/pull/6137#issuecomment-2258964204), although it doesn't show the diff- but you get the diff when updating `api.txt` (which that comment enforces).

But doing it before a PR sounds like a stretch goal we might wanna record. I've created [b/369361559](https://b.corp.google.com/issues/369361559) to track this future work.

File: firebase-dataconnect/firebase-dataconnect.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+  alias(libs.plugins.kotlinx.serialization)
+}
+
+firebaseLibrary {
```
[the remark this answers is not in the document]
Comment (daymxn) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts:27]: Hmm, I'm not sure where the `version` is defined in G3.

@rachelsaunders would you happen to have context on this?

The `version` is what we use in the headers for release notes, eg;
```md
### {{crashlytics}} NDK version 19.0.3 {: #crashlytics-ndk_v19-0-3}
```
In this case, `crashlytics-ndk` is the `version` we use internally.

File: firebase-dataconnect/firebase-dataconnect.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+      }
+    }
+  }
+}
+
+dependencies {
+  api("com.google.firebase:firebase-common:21.0.0")
+
+  api(libs.kotlinx.coroutines.core)
+  api(libs.kotlinx.serialization.core)
```
[the remark this answers is not in the document]
Comment (daymxn) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts:117]: I feel like this discussion has been had before, but I couldn't find a record of it.

Technically, they don't _need_ to be `api`. For example, we don't expose coroutines on the consumers classpath for `firebase-database`; even though we have methods that return/accept `Flow`. This works, because consumers are expected to have coroutines on their classpath anyways- as it's a requirement on the Kotlin side. This also avoids the issue of adding additional bloat to the SDK size for consumers that _don't_ use those APIs, since coroutines is a "bulky" library.

BUT, we _should_ do this. The guidance is- if it's public api, then `api` it.

Also [relevant discussion in chat](https://chat.google.com/room/AAAATnZMo34/AXiRStLAO_4/AXiRStLAO_4?cls=10).

@rlazo Any thoughts?

**Review by rachelsaunders — CHANGES_REQUESTED** [review]

File: firebase-dataconnect/firebase-dataconnect.gradle.kts (hunk 2 above)
Comment (daymxn) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts:193]: I've created [b/369383096](https://b.corp.google.com/issues/369383096) to track this.

File: build.gradle [hunk 12]
```diff
@@ -41,6 +41,7 @@ buildscript {
         classpath 'com.google.firebase:firebase-appdistribution-gradle:5.0.0'
         classpath 'com.google.firebase:firebase-crashlytics-gradle:2.9.5'
         classpath "com.diffplug.spotless:spotless-plugin-gradle:7.0.0.BETA1"
+        classpath 'io.javalin:javalin:6.1.4'
```
[the remark this answers is not in the document]
Comment (daymxn) [reply build.gradle]: I'm marking this as resolved since this change was reverted per our [other discussion](https://github.com/firebase/firebase-android-sdk/pull/6290#discussion_r1771710582)

File: firebase-dataconnect/firebase-dataconnect.gradle.kts [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+  testImplementation(libs.robolectric)
+  testImplementation(libs.truth)
+  testImplementation(libs.truth.liteproto.extension)
+  testImplementation(libs.kotlin.coroutines.test)
+  testImplementation(libs.kotlinx.serialization.json)
+
+  androidTestImplementation(project(":firebase-dataconnect:androidTestutil"))
+  androidTestImplementation(project(":firebase-dataconnect:connectors"))
+  androidTestImplementation(project(":firebase-dataconnect:testutil"))
+  androidTestImplementation("com.google.firebase:firebase-auth:22.3.1") {
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts]: The excludes have been removed.

File: firebase-dataconnect/RELEASING.md [hunk 14]
```diff
@@ -0,0 +1,247 @@
+# Release Firebase Data Connect Android SDK
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/RELEASING.md]: Done. RELEASING.md has been deleted (https://github.com/firebase/firebase-android-sdk/pull/6290/commits/6d3405fa11e6576be686e77c635bea97bf031fb8)

File: firebase-dataconnect/RELEASING.md [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+Last updated June 25, 2024 by @dconeybe
+
+This document outlines the steps for releasing _alpha_ versions of the Firebase Data Connect Android
+SDK. At the time of writing (June 25, 2024) the Data Connect SDK lives in a branch named
+`dataconnect`. Once the `dataconnect` branch is merged into the `master` branch, these steps will be
+obsolete and will need to be updated.
+
+### Set Release Date in FirebaseDataConnect.kt
+
+Open [FirebaseDataConnect.kt](https://github.com/firebase/firebase-android-sdk/blob/dataconnect/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/FirebaseDataConnect.kt)
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/RELEASING.md]: I plan to move the release notes from FirebaseDataConnect.kt into CHANGELOG.md in a follow-up PR.

File: firebase-dataconnect/firebase-dataconnect.gradle.kts (hunk 11 above)
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/firebase-dataconnect.gradle.kts]: serialization is used here: https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/FirebaseDataConnect.kt#L26-L28

https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/FirebaseDataConnect.kt#L272-L278

coroutines is used here: https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/QuerySubscription.kt#L19

https://github.com/firebase/firebase-android-sdk/blob/6d3405fa11e6576be686e77c635bea97bf031fb8/firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/QuerySubscription.kt#L52

File: firebase-dataconnect/CHANGELOG.md [hunk 16]
```diff
@@ -1,3 +1,5 @@
 # Unreleased
-* [feature] Add ability to specify SerializersModule when serializing.
 * [feature] initial beta release.
```
[the remark this answers is not in the document]
Comment (dconeybe) [reply firebase-dataconnect/CHANGELOG.md]: Public preview wording has been changed, as suggested: https://github.com/firebase/firebase-android-sdk/pull/6290/commits/9d47d2ca88edd670490b51d40c59ba90d867020c

**Review by daymxn — COMMENTED** [review]
+1 

Build files look fine to me. Still have pending work that needs @rachelsaunders input [here](https://github.com/firebase/firebase-android-sdk/pull/6290#discussion_r1773630509), but otherwise I defer to @rlazo for final approval.

Also, worth looking at the smoke test failure. Those don't usually flake.

**Review by rlazo — APPROVED** [review]

**Comment by dconeybe** [comment]
@rlazo I may need you to merge this PR fore because it won't let me because the cla presubmit is broken with such a large PR.

**Comment by dconeybe** [comment]
> +1
> 
> Build files look fine to me. Still have pending work that needs @rachelsaunders input [here](https://github.com/firebase/firebase-android-sdk/pull/6290#discussion_r1773630509), but otherwise I defer to @rlazo for final approval.
> 
> Also, worth looking at the smoke test failure. Those don't usually flake.

The most recent failure of the smoke test (at the time of writing) was due to this error:

> `This project's quota for this operation has been exceeded. [ Exceeded quota for verifying passwords. ]`

(http://shortn/_hOLZ5pldH3)

A re-run fixed it.

Also, I've opened 2 PRs to fix flaky tests that I came across: https://github.com/firebase/firebase-android-sdk/pull/6290 and https://github.com/firebase/firebase-android-sdk/pull/6291. I'll continue to monitor the flaky failures.

File: firebase-dataconnect/CHANGELOG.md [hunk 17]
```diff
@@ -0,0 +1,7 @@
+# Unreleased
+* [feature] Initial release of the Data Connect SDK (public preview). See
+  https://firebase.google.com/docs/data-connect/android-sdk for instructions on
+  getting started.
```
Comment (rachelsaunders) [inline firebase-dataconnect/CHANGELOG.md:4]:
```suggestion
* [feature] Initial release of the Data Connect SDK (public preview). Learn how to
  [get started](https://firebase.google.com/docs/data-connect/android-sdk)
  with the SDK in your app.
```

File: firebase-dataconnect/CHANGELOG.md [hunk 18]
```diff
@@ -0,0 +1,7 @@
...
+* [feature] Add ability to specify SerializersModule when serializing.
```
Comment (rachelsaunders) [inline firebase-dataconnect/CHANGELOG.md:5]:
```suggestion
* [feature] Added ability to specify `SerializersModule` when serializing.
```

File: firebase-dataconnect/CHANGELOG.md [hunk 19]
```diff
@@ -0,0 +1,7 @@
...
+* [feature] CallerSdkType added, to enable tracking generated SDK usage.
```
Comment (rachelsaunders) [inline firebase-dataconnect/CHANGELOG.md:6]:
```suggestion
* [feature] Added `CallerSdkType`, which enables tracking of the generated SDK usage.
```

**Review by rachelsaunders — CHANGES_REQUESTED** [review]
The info in this changelog gets transformed into release notes displayed in the docs website.  So it need to be markdown that renders appropriately.  No raw URLs.  Proper syntax, like backticks.  Start the entry whenever possible with a verb in the past tense.  Use proper brand names.

If you have the PR/Issue for each of these (except the "initial release" one), then please add that, too.

Thx!

**Comment by dconeybe** [comment]
@rachelsaunders 

> The info in this changelog gets transformed into release notes displayed in the docs website. So it need to be markdown that renders appropriately. No raw URLs. Proper syntax, like backticks. Start the entry whenever possible with a verb in the past tense. Use proper brand names.
> 
> If you have the PR/Issue for each of these (except the "initial release" one), then please add that, too.

Done. Please take a look at https://github.com/firebase/firebase-android-sdk/blob/dataconnect/firebase-dataconnect/CHANGELOG.md.

Also, do you know what values to specify for `name.set` and `versionName.set` in the `releaseNotes` section of [`firebase-dataconnect.gradle.kts`](https://github.com/firebase/firebase-android-sdk/blob/bcda6da5fbd81c2a7e2b6c241df26542383206ae/firebase-dataconnect/firebase-dataconnect.gradle.kts#L34-L35)? The values there are merely guesses on my part.

**Comment by rachelsaunders** [comment]
> @rachelsaunders
> 
> > The info in this changelog gets transformed into release notes displayed in the docs website. So it need to be markdown that renders appropriately. No raw URLs. Proper syntax, like backticks. Start the entry whenever possible with a verb in the past tense. Use proper brand names.
> > If you have the PR/Issue for each of these (except the "initial release" one), then please add that, too.
> 
> Done. Please take a look at https://github.com/firebase/firebase-android-sdk/blob/dataconnect/firebase-dataconnect/CHANGELOG.md.

Beautiful :-D

> 
> Also, do you know what values to specify for `name.set` and `versionName.set` in the `releaseNotes` section of [`firebase-dataconnect.gradle.kts`](https://github.com/firebase/firebase-android-sdk/blob/bcda6da5fbd81c2a7e2b6c241df26542383206ae/firebase-dataconnect/firebase-dataconnect.gradle.kts#L34-L35)? The values there are merely guesses on my part.

This would be something that @daymxn or @rlazo can help you with.

**Review by rachelsaunders — APPROVED** [review]

**Comment by dconeybe** [comment]
To see a snapshot of the `dataconnect` branch as it was when it was merged into `main` see https://github.com/dconeybe/dataconnect-android-sdk
